Provide a `/wopi/iop/open` endpoint

glpatcern opened the pull request:
To match https://github.com/cs3org/cs3apis/pull/75, this implements a new `/wopi/iop/open` endpoint with the following arguments:
Required headers:
  - `Authorization: Bearer` and the shared Reva/WOPI secret
  - `TokenHeader`: an x-access-token to serve as user identity towards Reva

Query parameters:
  - string viewmode: how the user should access the file - one of `VIEW_MODE_VIEW_ONLY`, `VIEW_MODE_READ_ONLY`, `VIEW_MODE_READ_WRITE`, defaults to `VIEW_MODE_READ_ONLY`
  - string username (optional): user's full display name, typically shown by the Office app
  - string filename OR fileid: the full path of the filename to be opened, or its fileid
  - string folderurl: the URL to come back to the containing folder for this file, typically shown by the Office app
  - string endpoint (optional): the storage endpoint to be used to look up the file or the storage id, in case of multi-instance underlying storage; defaults to 'default'

The former whitelist mechanism is dropped, given the double protection provided by the required headers.

LovisaLugnegard commented on src/wopiserver.py:
```diff
@@ -290,18 +292,23 @@ def cboxOpen():
             Wopi.log.info('msg="cboxOpen: remote error on generating token" client="%s" user="%s" ' \
-                          'friendlyname="%s" canedit="%s" endpoint="%s" reason="%s"' % \
-                          (req.remote_addr, userid, username, canedit, endpoint, e))
+                          'friendlyname="%s" mode="%s" endpoint="%s" reason="%s"' % \
```
Shouldn't this be called viewmode?

LovisaLugnegard commented on src/wopiserver.py:
```diff
@@ -311,6 +318,31 @@ def cboxOpen():
+  '''Returns a list of all currently opened files, for operations purposes only.
+  This call is protected by the same shared secret as the /wopi/iop/open call.'''
+  req = flask.request
+  # first check if the shared secret matches ours
```
Maybe this comment can be removed?

LovisaLugnegard commented on src/wopiserver.py:
```diff
@@ -311,6 +318,31 @@ def cboxOpen():
+                     'client="%s"' % req.remote_addr)
+    return 'Client not authorized', http.client.UNAUTHORIZED
+  # first convert the sets into lists, otherwise sets cannot be serialized in JSON format
+  jl = {}
```
What does jl stand for?

glpatcern replied:
json-able list. OK, that's too short, I renamed it to `jlist`. Again it's almost a mute variable, generated there and sent over in the next line.

LovisaLugnegard commented on src/wopiserver.py:
```diff
@@ -311,6 +318,31 @@ def cboxOpen():
+    return 'Client not authorized', http.client.UNAUTHORIZED
+  # first convert the sets into lists, otherwise sets cannot be serialized in JSON format
+  jl = {}
+  for f in list(Wopi.openfiles.keys()):
```
f=file?

glpatcern replied:
Yes. As usual (for me), mute variables like loop iterators are single-letter identifiers ;-)

LovisaLugnegard replied:
:) I don't agree here, but if you've a very strong opinion I'm fine with these single letters.

LovisaLugnegard commented on src/wopiutils.py:
```diff
@@ -19,14 +20,25 @@
 
-credentials = {}
 
+class ViewMode(IntEnum):
```
Why do you use intEnum instead of Enum?

glpatcern replied:
Very good question: python does not support ints in `Enum` types, and one has to use this derived type instead!

LovisaLugnegard replied:
Yes, but why do we use ints instead of "INVALID"= "INVALID", "VIEW_ONLY"="VIEW_ONLY" etc?

glpatcern replied:
I thought that the CS3API enum itself was defined with ints, and Reva would pass ints to the call. Can we have strings instead? if so, I'm happy to use them :-)

LovisaLugnegard replied:
I guess it depends on how we call the wopi server from reva. I don't think passing a string enum should be a problem.

glpatcern replied:
Well, happy to go for that then - and there's no eventuality of "nil" or invalid values as those would be rejected by the open call itself.
I'll change the `Enum` to match strings.

LovisaLugnegard commented on src/wopiutils.py:
```diff
@@ -95,19 +106,19 @@ def generateAccessToken(userid, fileid, canedit, username, folderurl, endpoint):
                        'extlock': locked, 'folderurl': folderurl, 'exp': exptime, 'endpoint': endpoint}, \
                       _ctx['wopi'].wopisecret, algorithm='HS256').decode('UTF-8')
-  _ctx['log'].info('msg="Access token generated" userid="%s" canedit="%r" filename="%s" inode="%s" ' \
+  _ctx['log'].info('msg="Access token generated" userid="%s" mode="%s" filename="%s" inode="%s" ' \
```
Maybe have `viewmode` in the log as well?

glpatcern replied:
I've shortened it after a first go as `viewmode` because this is how it is logged, and it's already pretty verbose:
```
2020-06-16T09:48:13 flask.app[26886] INFO     msg="Access token generated" userid="...:..." mode="ViewMode.READ_WRITE" filename="/eos/user/.../Example.odt" inode="1003128628997504" mtime="1589374408" folderurl="foo" expiration="1592380093" token="Z65sIVe-1DAZf9W2KIEc"
```

LovisaLugnegard replied:
Yes, it's verbose. But if I were to debug anything with `viewmode` I'd expect to be able to ctrl+f "viewmode" to find everything needed instead of having to figure out that I actually need to ctrl+f "mode". Ofc I can look in the code, but that takes much more time.

glpatcern replied:
...which would work as the value is `ViewMode.xxx`, thanks to the nice python representation of enums...

LovisaLugnegard replied:
But what happens if the problem would be that the value is nil?

glpatcern replied:
After the last commit, this option cannot happen - any value different from the 3 valid ones is discarded with a HTTP BAD_REQUEST.

LovisaLugnegard commented on tools/wopiopen.py:
```diff
@@ -11,21 +11,24 @@
 def usage(exitcode):
   '''Prints usage'''
-  print('Usage : ' + sys.argv[0] + ' [-h|--help] <filename> <uid> <gid>')
+  print('Usage : ' + sys.argv[0] + ' [-h|--help] [-m|--mode 1|2|3] <filename> <uid> <gid>')
```
mode == viewmode?

glpatcern replied:
Yes, but v == verbose. Anyway this `wopiopen` CLI is a kind of internal command for us to quickly test, and soon it is to be replaced by (your) Reva open CLI.

LovisaLugnegard replied:
As for the other comment, I think we should be consistent even for the internal namings and loggins.

glpatcern replied:
OK, I realize that in this command there's no (any more) references to `verbose` (which I typically use for verbose output/logging), so let's use `viewmode` and drop verbose then.

LovisaLugnegard commented on tools/wopiopen.py:
```diff
@@ -11,21 +11,24 @@
   print(e)
   usage(1)
 verbose = False
+mode = 3   # 3 is full access, 2 for read-only, 1 for view-only.
```
I think we could have ordinary Enums and check for them. Then we won't need this comment.

glpatcern replied:
Indeed that was a leftover

LovisaLugnegard commented on tools/wopiopen.py:
```diff
@@ -11,21 +11,24 @@
   usage(1)
 verbose = False
+mode = 3   # 3 is full access, 2 for read-only, 1 for view-only.
 for f, v in options:
```
What does f stand for?

glpatcern replied:
flag and value. This is boiler-plate code that I copy over every time I have to parse some command line arguments...

LovisaLugnegard replied:
These single letters 🤯 I'd prefer if we wrote flag and value to make to code easier to understand by a quick look. But I'm also all for democracy so if I'm the minority, let's continue with this.

LovisaLugnegard commented on tools/wopiopen.py:
```diff
@@ -65,10 +68,10 @@ def usage(exitcode):
                        headers={'Authorization': 'Bearer ' + iopsecret},
                        params={'ruid': ruid, 'rgid': rgid, 'filename': filename, 'endpoint': endpoint,
-                               'canedit': 'True', 'username': 'Operator', 'folderurl': 'foo'})
+                               'viewmode': mode, 'username': 'Operator', 'folderurl': 'foo'})
```
I think we should call viewmode "viewmode" everywhere to be consistent.

glpatcern commented:
A general comment: most of the code did not really change and was only moved around. Apart addressing the remarks, we need to discuss if we want to drop the `allowedclients` logic as with Reva the authentication relies on both a shared secret and the `TokenHeader`, both non-guessable.
